pkg/pillar/hypervisor/kubevirt.go

```diff
@@ -0,0 +1,1421 @@
+			r.MaxUsedMemory = r.UsedMemory
+		}
+		if _, ok := ctx.prevDomainMetric[n]; !ok {
+			r.MaxUsedMemory = r.UsedMemory
+			ctx.prevDomainMetric[n] = r
+		} else {
+			if ctx.prevDomainMetric[n].MaxUsedMemory > r.UsedMemory {
+				r.MaxUsedMemory = ctx.prevDomainMetric[n].MaxUsedMemory
+			} else if ctx.prevDomainMetric[n].MaxUsedMemory < r.UsedMemory {
+				r.MaxUsedMemory = r.UsedMemory
+			}
+			ctx.prevDomainMetric[n] = r
+		}
+		res[n] = r
+	}
+
+	hasEmptyRes := len(ctx.vmiList) - len(res)
+	if hasEmptyRes > 0 {
+		// check and get the kubernetes pod's metrics
+		checkPodMetrics(ctx, res, hasEmptyRes)
+	}
+	logrus.Debugf("GetDomsCPUMem: %d VMs: %+v, podnum %d", len(ctx.vmiList), res, hasEmptyRes)
+	return res, nil
+}
+
+func getVirtHandlerIPAddr(ctx *kubevirtContext) (string, error) {
+	if ctx.virthandlerIPAddr != "" {
+		return ctx.virthandlerIPAddr, nil
+	}
+	clientSet, err := kubeapi.GetClientSet()
+	if err != nil {
+		return "", err
+	}
+
+	pods, err := clientSet.CoreV1().Pods("kubevirt").List(context.Background(),
+		metav1.ListOptions{
+			LabelSelector: "kubevirt.io=virt-handler",
+		})
+	if err != nil {
+		return "", err
+	}
+
+	var vmiPod *k8sv1.Pod
+	for _, pod := range pods.Items {
+		if strings.HasPrefix(pod.ObjectMeta.Name, "virt-handler-") {
+			vmiPod = &pod
+			break
+		}
+	}
+
+	if vmiPod == nil {
+		return "", fmt.Errorf("can not find virt-handler pod")
+	}
+	ctx.virthandlerIPAddr = vmiPod.Status.PodIP
+	return ctx.virthandlerIPAddr, nil
+}
+
+func getVMIName(metricStr string) string {
+	nameStr := strings.SplitN(metricStr, ",name=\"", 2)
+	if len(nameStr) != 2 {
+		fmt.Printf("get name failed, string %s\n", metricStr)
+		return ""
+	}
+	nameStr2 := strings.SplitN(nameStr[1], "\",", 2)
+	if len(nameStr2) != 2 {
+		fmt.Printf("get name2 failed, string %s\n", metricStr)
+		return ""
+	}
+	return nameStr2[0]
+}
+
+func fillMetrics(res map[string]types.DomainMetric, domainName, metricName string, value interface{}) {
+	if _, ok := res[domainName]; !ok {
+		logrus.Infof("fillMetrics, vmiName %s not in map", domainName)
+		return
+	}
+
+	r := res[domainName]
+	BytesInMegabyte := uint32(1024 * 1024)
+	switch metricName {
+	// add all the cpus to be Total, seconds should be from VM startup time
+	case "kubevirt_vmi_cpu_system_usage_seconds":
+	case "kubevirt_vmi_cpu_usage_seconds":
+	case "kubevirt_vmi_cpu_user_usage_seconds":
+		cpuNs := assignToInt64(value) * int64(time.Second)
+		r.CPUTotalNs = r.CPUTotalNs + uint64(cpuNs)
+	case "kubevirt_vmi_memory_usable_bytes":
+		r.AvailableMemory = uint32(assignToInt64(value)) / BytesInMegabyte
+	case "kubevirt_vmi_memory_domain_bytes_total":
+		r.AllocatedMB = uint32(assignToInt64(value)) / BytesInMegabyte
+	case "kubevirt_vmi_memory_available_bytes": // save this temp for later
+		r.UsedMemory = uint32(assignToInt64(value)) / BytesInMegabyte
+	//case "kubevirt_vmi_memory_resident_bytes":
+	//	r.UsedMemory = uint32(assignToInt64(value)) / BytesInMegabyte
```

Comment:
Why is it commented out?

Reply:
we can remove this, this was used to experiment on which one makes sense.

Reply:
ack, will be in next commit